### PR TITLE
Support for latency-based load balancing across multiple key servers

### PR DESCRIPTION
**Pull request by jacobhaven**
Initial stab at latency based load balancing across multiple server IPs using a priority queue of server Addrs.

### CODE REVIEW

File: client/remote.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+			if err := conn.Ping(nil); err != nil {
+				log.Infof("Connection to %s failed: %v", s, err)
+				return
+			}
+			duration := time.Since(start)
+
+			r.Lock()
+			s.latency.Update(duration)
+			if s.index >= 0 {
+				heap.Fix(r, s.index)
```
Comment (lziest) [inline client/remote.go]: would this cause out-of-bound error? s is not longer in r?

Comment (jacobhaven) [inline client/remote.go]: It should be in `r` (it's only removed removed while the lock is helf), but if it isn't `s.index` is set to `-1` (thus the `>=0` check)

Comment (lziest) [inline client/remote.go]: got it!

**Comment by lziest** [comment]
CI error is git error :(
LGTM.

File: client/remote.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+func (sh serverHeap) Swap(i, j int) {
+	sh[i], sh[j] = sh[j], sh[i]
+	sh[i].index = i
+	sh[j].index = j
+}
+
+func (sh serverHeap) Less(i, j int) bool {
+	return sh[j].conn == nil || sh[i].latency.RateMean() < sh[j].latency.RateMean()
```
Comment (grittygrease) [inline client/remote.go]: Is RateMean the right metric to be using? Recent requests should be more highly weighted than old requests.

File: client/client.go [hunk 3]
```diff
@@ -116,26 +81,10 @@ func (c *Client) DialAny(ski gokeyless.SKI) (*gokeyless.Conn, error) {
...
-			}
-			c.m.Lock()
-			if c.conns[server] == conn {
-				delete(c.conns, server)
-			}
-			c.m.Unlock()
-		}
-	}
-
 	// choose from possible servers at random until a connection can be established.
```
Comment (jacobhaven) [inline client/client.go]: This is a bit of a punt. Really there should only be one `Remote` registered. We could enforce that by having `registerSKI` append to the current remote (creating a `group` if necessary).

File: client/client.go [hunk 4]
```diff
@@ -197,31 +170,75 @@ func (c *Client) ValidServer(server string) (err error) {
...
-// registerSKI associates the SKI of a public key with a particular keyserver.
-func (c *Client) registerSKI(server string, ski gokeyless.SKI) error {
-	if server == "" {
-		server = c.DefaultServer
+// AddRemote adds a remote for the given SKI, optionally registering it
+// under the given server name.
+func (c *Client) AddRemote(server string, r Remote, ski gokeyless.SKI) *PrivateKey {
+	c.m.Lock()
+	defer c.m.Unlock()
+
```
Comment (jacobhaven) [inline client/client.go]: Make sure this logic all makes the most sense.

File: client/client.go [hunk 5]
```diff
@@ -77,151 +75,125 @@ func NewClientFromFile(certFile, keyFile, caFile string) (*Client, error) {
...
+// AddRemote adds a remote for the given SKI, optionally registering it
+// under the given server name.
+func (c *Client) AddRemote(server string, r Remote, ski gokeyless.SKI) *PrivateKey {
+	c.m.Lock()
+	defer c.m.Unlock()
 
-	var ips []net.IP
-	if ips, err = net.LookupIP(host); err != nil {
-		return
+	if server != "" {
```
Comment (jacobhaven) [inline client/client.go]: I could use some eyes on all the cached `c.servers` and `c.remotes` logic to make sure it make sense.

**Comment by lziest** [comment]
lgtm again!

**Pull request merged**